crossbeam-epoch/src/atomic.rs

```diff
@@ -103,15 +103,20 @@ fn low_bits<T>() -> usize {
 ///
 /// `tag` is truncated to fit into the unused bits of the pointer to `T`.
 #[inline]
-fn data_with_tag<T>(data: usize, tag: usize) -> usize {
-    (data & !low_bits::<T>()) | (tag & low_bits::<T>())
+fn data_with_tag<T>(data: *mut T, tag: usize) -> *mut T {
+    // transmute preserves pointer provenance
+    let data: usize = core::mem::transmute(data);
+    core::mem::transmute((data & !low_bits::<T>()) | (tag & low_bits::<T>()))
```

Comment:
If the transmute preserved provenance, doing integer operations on the pointer will still kill it unless we do really awful things (things that we used to do but that I very happily removed when integer-ptr-casts got properly supported).

Reply:
Hmm, interesting. Is there any way to support provenance through tagged pointers then?

Reply:
See `ptr_int_arithmetic` removed [in this commit](https://github.com/rust-lang/miri/commit/82da68c5caca6b534d77c01cbcf9d1ed0a6ef496) for the awful things we used to do.

Reply:
I opened https://github.com/rust-lang/miri/issues/1318 to track the Miri-side issue (and to avoid having this discussion in 4 threads in parallel^^).